From 3b5149e4cf7b4209705fd9d671203a71ffd294c1 Mon Sep 17 00:00:00 2001
From: Wilfred Mallawa <wilfred.mallawa@wdc.com>
Date: Wed, 26 Jul 2023 16:25:58 +1000
Subject: [PATCH 3/3] PCI: rockchip-host: Add a shutdown driver handler

Add a shutdown handler with only the functionality to assert
the PERST signal to indicate a reset.

Signed-off-by: Wilfred Mallawa <wilfred.mallawa@wdc.com>
---
 drivers/pci/controller/pcie-rockchip-host.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/pci/controller/pcie-rockchip-host.c b/drivers/pci/controller/pcie-rockchip-host.c
index 07aeee132ae3..839e371e0f67 100644
--- a/drivers/pci/controller/pcie-rockchip-host.c
+++ b/drivers/pci/controller/pcie-rockchip-host.c
@@ -1023,6 +1023,18 @@ static int rockchip_pcie_probe(struct platform_device *pdev)
 	return err;
 }
 
+static void rockchip_pcie_shutdown(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct rockchip_pcie *rockchip = dev_get_drvdata(dev);
+	int err;
+
+	/* Assert PERST to indicate link going down */
+	err = rockchip_pcie_host_issue_perst(rockchip);
+	if (err)
+		dev_err(dev, "Failed to issue PERST");
+}
+
 static void rockchip_pcie_remove(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -1063,6 +1075,7 @@ static struct platform_driver rockchip_pcie_driver = {
 		.pm = &rockchip_pcie_pm_ops,
 	},
 	.probe = rockchip_pcie_probe,
+	.shutdown	= rockchip_pcie_shutdown,
 	.remove_new = rockchip_pcie_remove,
 };
 module_platform_driver(rockchip_pcie_driver);
-- 
2.41.0

